OHFJIRA-52 list of exposed webservices

kkovarik opened the pull request:
### ABOUT 
Overview of exposed Webservices. See related JIRA task for more info: https://openhubframework.atlassian.net/browse/OHFJIRA-52

### CHANGES
* added new REST endpoint GET /api/ws with simple test. It does reuse WsdlRegistry functionality.

kkovarik committed:
OHFJIRA-52 list of exposed webservices

pjuza commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/ws/rest/WsController.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@RestController
+@RequestMapping(value = WsController.REST_URI)
+public class WsController extends AbstractOhfController {
```
Maybe WsdlController would be better ...

kkovarik replied:
And where to expose it, directly under /api/wsdl?

No problem with that, I do not have any strict opinion on this, my initial thinking was like - ok, what is wsdl? definition of service, resource itself desired by consumer is the webservice.

pjuza commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/ws/rpc/WsInfoRpc.java:
```diff
@@ -0,0 +1,61 @@
+ * @author Karel Kovarik
+ */
+@XmlRootElement
+public class WsInfoRpc {
```
Same as controller - wouldn't be better to use "wsdl" prefix, isn't it more descriptive?

kkovarik replied:
Possibly, as in previous comment, if it is the wsdl resource this endpoint should list, then yes, I will change it...

pjuza replied:
Yes, it's overview of input web service definitions (wsdl).
We will add overview of REST services in the next version ...

pjuza reviewed (changes requested):
Is there API documentation in Apiary? Also mockup diagram for GUI developers in Balsamiq would be nice.

kkovarik commented:
I apologize, the links to drafts of mockup and API documentation were mentioned only in the JIRA task:

Api draft: http://docs.openhubframework.apiary.io/#reference/0/openhub-web-services/gets-list-of-exposed-webservices
Mockup draft: https://openwise.mybalsamiq.com/projects/openhub/WS%20overview

pjuza commented:
Please change URI: /api/ws -> /web/admin/api/ws

If you accept change from ws to wsdl then please change also names in API.

hanusto commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/ws/rest/WsController.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Base REST uri for the controller.
+     */
+    public static final String REST_URI = BASE_PATH + "/ws";
```
And WSDL for endpoint URL too. Maybe my recommendation is `/api/services/wsdl`

kkovarik replied:
Done.

hanusto commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/ws/rest/WsController.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Uri part on which wsdl is exposed.
+     */
+    private static final String WS_URI = "/ws/";
```
I dont know if the WS context can be configured or not. If yes, this should be the same.

kkovarik replied:
Cannot, however changed it to the constant used to actually configure the context (RouteConstants.WS_URI_PREFIX).

hanusto commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/ws/rest/WsController.java:
```diff
@@ -0,0 +1,101 @@
+     * @param request the input httpServletRequest.
+     * @return list of WsInfoRpc domain objects.
+     */
+    @GetMapping(produces = {"application/xml", "application/json"})
```
great

hanusto reviewed (changes requested):
I recommend to rename WS -> WSDL and whole endpoint should be in /api/services context. After that I agree with that.

kkovarik commented:
Thanks for the review, updated endpoint to **/api/services/wsd**l and refactored to WsdlController & WsdlInfoRpc etc. feel free to check again.

API documentation updated: http://docs.openhubframework.apiary.io/#reference/0/openhub-web-services, now I am using collection wrapping as it is introduced in PR#53: https://github.com/OpenWiseSolutions/openhub-framework/pull/53

Will wait until PR#53 is merged, then rebase this branch to use the common CollectionWrapper as well...

hanusto commented on web-admin/src/main/java/org/openhubframework/openhub/admin/web/services/wsdl/rest/WsdlController.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Get list of exposed web services.
+     * @param request the input httpServletRequest.
+     * @return list of WsInfoRpc domain objects wrapped in CollectionWrapper.
```
WsInfoRpc -> WsdlInfoRpc, or use {@link WsInfoRpc}. This was refactored automatically. The same for {@link CollectionWrapper}.

kkovarik replied:
Thanks, missed that, fixed.

hanusto reviewed (approved):
Now it is perfect. Only one improvement introduced -> javadoc. PR https://github.com/OpenWiseSolutions/openhub-framework/pull/53 merged so you can rebase it with requested changes.

hanusto commented:
PR #53 is merged, so you can rebase that.

kkovarik commented:
@hanusto thanks for fast merge, rebased on current develop.
